src/vanilla/utils/proxyMap.ts

```diff
@@ -1,111 +1,166 @@
-import { proxy } from '../../vanilla.ts'
+import { proxy, unstable_getInternalStates } from '../../vanilla.ts'
 
-type KeyValRecord<K, V> = [key: K, value: V]
+const { proxyStateMap, snapCache } = unstable_getInternalStates()
+const maybeProxify = (x: any) => (typeof x === 'object' ? proxy({ x }).x : x)
+const isProxy = (x: any) => proxyStateMap.has(x)
 
-type InternalProxyMap<K, V> = Map<K, V> & {
-  data: KeyValRecord<K, V>[]
-  toJSON: object
+type InternalProxyObject<K, V> = Map<K, V> & {
+  data: Array<K | V>
+  index: number
+  toJSON: () => Map<K, V>
 }
 
-/**
- * proxyMap
- *
- * This is to create a proxy which mimic the native Map behavior.
- * The API is the same as Map API
- *
- * @example
- * import { proxyMap } from 'valtio/utils'
- * const state = proxyMap([["key", "value"]])
- *
- * //can be used inside a proxy as well
- * const state = proxy({
- *   count: 1,
- *   map: proxyMap()
- * })
- *
- * // When using an object as a key, you can wrap it with `ref` so it's not proxied
- * // this is useful if you want to preserve the key equality
- * import { ref } from 'valtio'
- *
- * const key = ref({})
- * state.set(key, "value")
- * state.get(key) //value
- *
- * const key = {}
- * state.set(key, "value")
- * state.get(key) //undefined
- */
-export function proxyMap<K, V>(entries?: Iterable<readonly [K, V]> | null) {
-  const map: InternalProxyMap<K, V> = proxy({
-    data: Array.from(entries || []) as KeyValRecord<K, V>[],
-    has(key) {
-      return this.data.some((p) => p[0] === key)
+export function proxyMap<K, V>(entries?: Iterable<[K, V]> | undefined | null) {
+  const initialData: Array<K | V> = []
+  let initialIndex = 0
+  const indexMap = new Map<K, number>()
+
+  const snapMapCache = new WeakMap<object, Map<K, number>>()
+  const registerSnapMap = () => {
+    const cache = snapCache.get(vObject)
+    const latestSnap = cache?.[1]
+    if (latestSnap && !snapMapCache.has(latestSnap)) {
+      const clonedMap = new Map(indexMap)
+      snapMapCache.set(latestSnap, clonedMap)
+      return true
+    }
+    return false
+  }
+  const getSnapMap = (x: any) => snapMapCache.get(x)
```

Comment:
It feels a bit repetitive, so we can make it like:
```suggestion
  const getMapForThis = (x: any) => snapMapCache.get(x) || indexMap
```
I'm not sure about the naming.